content/6_Advanced/Critical.mdx

```diff
@@ -1,26 +1,526 @@
+		if (dsu_parent[dsu_parent[x]] != -1) {
+			int parent_res = find(dsu_parent[x]);
+			if (visit_time[parent_res] < visit_time[dsu_min[x]]) {
+				dsu_min[x] = parent_res;
+			}
+			dsu_parent[x] = dsu_parent[dsu_parent[x]];
+		}
+		assert(dsu_min[x] != INT32_MAX);
+		return dsu_min[x];
+	};
+
+	vector<int> sdom(city_num, -1);
+	for (int i = visit_order.size() - 1; i > 0; i--) {
+		int c = visit_order[i];
+		// Iterate over all nodes with a directed edge to c
+		for (int from : rev_adj[c]) {
+			int find_res = find(from);
+			// Take the node with the earliest visit time in the traversal
+			if (sdom[c] == -1 || visit_time[find_res] < visit_time[sdom[c]]) {
+				sdom[c] = find_res;
+			}
+		}
+
+		// Link c to its parent in the DFS tree
+		dsu_parent[c] = visit_parent[c];
+		dsu_min[c] = sdom[c];
+	}
+
+	// Initialize the values in the tree
+	vector<std::pair<int, int>> sdom_times(city_num, {INT32_MAX, -1});
+	for (int c : visit_order) {
+		if (c != 0) {
+			// The tree takes the minimum of these pairs, but what we
+			// really want is the node itself- therefore, we store both
+			sdom_times[c] = {visit_time[sdom[c]], c};
+		}
+	}
+	Tree tree(visit_parent, sdom_times);
+
+	vector<int> rdom(city_num, -1);
+	for (int c : visit_order) {
+		if (c != 0) {
+			// Use the definition of the rdom
+			rdom[c] = tree.min_val(sdom[c], c).second;
+		}
+	}
+
+	// Use the properties previously given to compute the idom
+	vector<int> idom(city_num, -1);
+	for (int c : visit_order) {
+		if (c != 0) { idom[c] = rdom[c] == c ? sdom[c] : idom[rdom[c]]; }
+	}
+
+	// Trace the idom tree from the finish node to the start node,
+	// finding all dominators along the way
+	vector<int> critical{0};
+	int at = city_num - 1;
+	while (at != 0) {
+		critical.push_back(at);
+		at = idom[at];
+	}
+
+	std::sort(critical.begin(), critical.end());
+
+	cout << critical.size() << '\n';
+	for (int i = 0; i < critical.size() - 1; i++) {
+		cout << critical[i] + 1 << ' ';
+	}
+	cout << critical.back() + 1 << endl;
+}
+```
+
+</CPPSection>
+</LanguageSection>
 
 ## Cycles
 
-- GP of Wroclaw 2020 H
-- [The Meeting Place Cannot be Changed](https://codeforces.com/problemset/problem/982/F)
-- USACO Camp - Acyclic Graphs
+<FocusProblem problem="cycles-critical" />
+
+The problem asks as to find the an intersection node of all cycles, if such exists. Firstly, let's remove all the nodes that don't belong to any cycle. To achieve this,
+recursivelly remove the nodes without any outgoing edge; i.e. after finishing with a node check it's parents for no outgoing edges. Now our graph is formed out of cycles.
+Assuming that the intersection of all cycles is not empty, start start reducing the cycles node by node. The last node standing should be the intersection.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <functional>
+#include <iostream>
+#include <unordered_set>
+#include <vector>
+
+using namespace std;
+
+const int NMAX = 1e5;
+
+unordered_set<int> in[NMAX], out[NMAX];
```

Comment:
does this actually tle w/ stock set

Reply:
Yes it does get tle

Reply:
Take a look:
https://codeforces.com/contest/982/submission/261065993